Adding ingest docs information for k8s (#630)
elastic/ingest-docs: docs/en/ingest-management/elastic-agent/ingest-pipeline-in-Kubernetes.asciidoc

[[ingest-pipeline-in-Kubernetes]]
= Using Custom Ingest Pipeline with {k8s} Integration

> This tutorial explains how to add a custom ingest pipeline to {k8s} Integration in order to add specific metadata fields for deployments and cronjobs of pods.

Reviewer: ```suggestion
This tutorial explains how to add a custom ingest pipeline to a {k8s} Integration in order to add specific metadata fields for deployments and cronjobs of pods.
```

> Custom pipelines can be used to add custom data processing, like adding fields, obfuscate sensitive information, and more. Find more information in [data-streams-pipeline-tutorial](https://www.elastic.co/guide/en/fleet/master/data-streams-pipeline-tutorial.html) on how to use custom ingest pipelines.

Reviewer: ```suggestion
Custom pipelines can be used to add custom data processing, like adding fields, obfuscating sensitive information, and more. Find more information in our tutorial about <<ingest-pipeline-in-Kubernetes,tutorial,transforming data with custom ingest pipelines>>.
```

== Metadata Enrichment for Kubernetes

> The https://docs.elastic.co/en/integrations/kubernetes[{k8s} Integration] is used to collect logs and metrics from Kubernetes clusters with {agent}. During the collection, integration enhances the collected information with extra useful information that can help users to correlate information and different Kubernetes assets with each other. All this additional information that are added on top of collected information like labels, annotations, ancestor names of Kubernetes assets etc. are called metadata.

Reviewer: ```suggestion
The https://docs.elastic.co/en/integrations/kubernetes[{k8s} Integration] is used to collect logs and metrics from Kubernetes clusters with {agent}. During the collection, the integration enhances the collected information with extra useful information that users can correlate with different Kubernetes assets. This additional information added on top of collected data, such as labels, annotations, ancestor names of Kubernetes assets, and others, are called metadata.
```
Just rephrasing slightly since `information` appears several times.

The https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html[{k8s} Provider] offers the `add_resource_metadata` option to configure the metadata enrichment options.

> For {agent} versions >[8.10.4], the default configuration for metadata enrichment is `add_resource_metadata.deployment=false` and `add_resource_metadata.cronjob=false`. This means that pods that created from replicasets that belong to specific deployments would not be enriched with `kubernetes.deployment.name`. Additionally, pods that are created from jobs that belong to specific cronjobs, would not be enriched with `kubernetes.cronjob.name`.

Reviewer: ```suggestion
For {agent} versions >[8.10.4], the default configuration for metadata enrichment is `add_resource_metadata.deployment=false` and `add_resource_metadata.cronjob=false`. This means that pods that are created from replicasets that belong to specific deployments would not be enriched with `kubernetes.deployment.name`. Additionally, pods that are created from jobs that belong to specific cronjobs, would not be enriched with `kubernetes.cronjob.name`.
```

**Kubernetes Integration Policy > Collect Kubernetes metrics from Kube-state-metrics > Kubernetes Pod Metrics**

--
[role="screenshot"]

> image::images/add_resource_metadata.png[Configure add_resource_metadata]

Reviewer: It looks like this image is missing. Adding it in might fix the failing docs CI check.

Author: My mistake added this

--

Eg. Enabling the enrichment through `add_resource_metadata` in Managed {agent} Policy

> **Note:** Enabling deployment and cronjob metadata enrichment leads to increase of {agent}'s memory consumption. Elastic agent utilizes local cache in order to keep records of the {k8s} assets being discovered.

== Add deployment and cronjob for {k8s} pods through ingest pipelines

> Alternative to keeping the feature on and use more memory resources for {agent}, users can make use of inegst pipelines to add the missing fields of `kubernetes.deployment.name` and `kubernetes.cronjob.name`.

Reviewer: ```suggestion
As an alternative to keeping the feature enabled and using more memory resources for {agent}, users can make use of ingest pipelines to add the missing fields of `kubernetes.deployment.name` and `kubernetes.cronjob.name`.
```

> Following article https://www.elastic.co/guide/en/fleet/master/data-streams-pipeline-tutorial.html[data-streams-pipeline-tutorial], navigate to `state_pod` datastream under: **Kubernetes Integration Policy > Collect Kubernetes metrics from Kube-state-metrics > Kubernetes Pod Metrics**

Reviewer: ```suggestion
Following the <<ingest-pipeline-in-Kubernetes,transforming data with custom ingest pipelines>> tutorial, navigate to `state_pod` datastream under: **Kubernetes Integration Policy > Collect Kubernetes metrics from Kube-state-metrics > Kubernetes Pod Metrics**.
```

Create the following custom ingest pipeline with two processors:
--
[role="screenshot"]
image::images/ingest_pipeline_custom_k8s.png[Custom ingest pipeline]
--

=== Processor for Deployment

--
[role="screenshot"]
image::images/gsub_deployment.png[Gsub Processor for deployment]
--

=== Processor for Cronjob

--
[role="screenshot"]
image::images/gsub_cronjob.png[Gsub Processor for cronjob]
--

The final `metrics-kubernetes.state_pod@custom` ingest pipeline:

[source,json]
----
[
  {
    "gsub": {
      "field": "kubernetes.replicaset.name",
      "pattern": "(?:.(?!-))+$",
      "replacement": "",
      "target_field": "kubernetes.deployment.name",
      "ignore_missing": true,
      "ignore_failure": true
    }
  },
  {
    "gsub": {
      "field": "kubernetes.job.name",
      "pattern": "(?:.(?!-))+$",
      "replacement": "",
      "target_field": "kubernetes.cronjob.name",
      "ignore_missing": true,
      "ignore_failure": true
    }
  }
]
----

> > **Note**: The ingest-pipeline does not check for the actual existence of deployment and cronjob ancestor, only adds the specific values.

Reviewer: ```suggestion
> **Note**: The ingest pipeline does not check for the actual existence of a deployment and cronjob ancestor, it only adds the specific values.
```
